openforcefield/typing/engines/smirnoff/parameters.py

```diff
@@ -471,19 +858,20 @@ def to_dict(self, discard_cosmetic_attributes=False):
 
         """
         # Make a list of all attribs that should be included in the
-        # returned dict (call list() to make a copy)
-        attribs_to_return = list(self._REQUIRED_SPEC_ATTRIBS)
-        attribs_to_return += [opt_attrib for opt_attrib in self._OPTIONAL_SPEC_ATTRIBS if hasattr(self, opt_attrib)]
+        # returned dict (call list() to make a copy). We discard
```

Comment:
~I think this is a behavior change from previous versions. Off the top of my head, I think it should only affect `(Im)ProperTorsion` `idivf` values (which are set to `auto` if not specified), but we should note it in the release notes.~

Reply:
Actually, we will want the _opposite_ behavior when we work with `ParameterHandler` -- That is, if the input didn't specify an `Electrostatics` `cutoff` value, we absolutely want to write it out explicitly. To keep things consistent, I think we should reverse this decision, and have the toolkit write out every default value that was assumed.

Reply:
I missed the first case, and I agree with the second. I'll change this, thank you! Let's make sure to tell @SimonBoothroyd as this will change which attributes are exported in the forcefield tags and thus the hash that is constructing.

Reply:
Good call. I just spoke with @SimonBoothroyd about this. Basically, in his hash/equality comparisons, _False Negatives are OK_ (they may degrade performance a little, but won't trigger a new simulation to be run or bad science to happen). So he's now aware of this toolkit behavior and will design around it.

Reply:
How about this: 4ac9e06. Basically, I'm not serializing attributes that are set to a default value only if their default value is `None`. This means we are discarding attributes like `id` and `parent_id` when they are not set, but catching attributes like `cutoff`.

Reply:
I like this. So basically, the difference between a required-attribute-with-a-default and an optional attribute is whether `default=val` or `default=None`. This seems like a good decision to me.

Reply:
More or less. They're both optional attributes when we read the forcefield, meaning that we don't have to specify them in the constructor, only we export only non-trivial defaults. The choice is just to avoid having to export a bunch of `parent_id=null` attributes for each tag.